File: Algorithms/Python/0116/0116.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if root == None:
            return None
        p, q = root.left, root.right
        while p != None and q != None:
            p.next = q
            p, q = p.right, q.left
        self.connect(root.left)
        self.connect(root.right)
        return root

File: Algorithms/C++/0116/0116.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if (root == nullptr) {
            return nullptr;
        }
        Node* p = root->left;
        Node* q = root->right;
        while (p != nullptr && q!= nullptr) {
            p->next = q;
            p = p->right;
            q = q->left;
        }
        connect(root->left);
        connect(root->right);
        return root;
    }
};
